primitives/xcm/src/lib.rs

```diff
@@ -344,3 +353,80 @@ impl<
 		}
 	}
 }
+
+// Struct that converts a given MultiLocation into a 20 bytes account id by hashing
+// with blake2_256 and taking the first 20 bytes
+pub struct Account20Hash<AccountId>(PhantomData<AccountId>);
+impl<AccountId: From<[u8; 20]> + Into<[u8; 20]> + Clone> Convert<MultiLocation, AccountId>
+	for Account20Hash<AccountId>
+{
+	fn convert_ref(location: impl Borrow<MultiLocation>) -> Result<AccountId, ()> {
+		let hash: [u8; 32] = ("multiloc", location.borrow())
+			.borrow()
+			.using_encoded(blake2_256);
+		let mut account_id = [0u8; 20];
+		account_id.copy_from_slice(&hash[0..20]);
+		Ok(account_id.into())
+	}
+
+	fn reverse_ref(_: impl Borrow<AccountId>) -> Result<MultiLocation, ()> {
+		Err(())
+	}
+}
+
+/// Allows execution from `origin` if it is contained in `T` (i.e. `T::Contains(origin)`) taking
+/// payments into account and if it starts with DescendOrigin.
+///
+/// Only allows for `DescendOrigin` + `WithdrawAsset`, + `BuyExecution`
+pub struct AllowDescendOriginFromLocal<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowDescendOriginFromLocal<T> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		_weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowTopLevelPaidExecutionFromLocal origin:
+			{:?}, message: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, message, max_weight, _weight_credit,
+		);
+		ensure!(T::contains(origin), ());
+		let mut iter = message.0.iter_mut();
+		let i = iter.next().ok_or(())?;
+
+		// Make sure the first instruction is DescendOrigin
+		match i {
+			DescendOrigin(..) => (),
+			_ => return Err(()),
+		}
+
+		// Then WithdrawAsset
+		let mut i = iter.next().ok_or(())?;
+		match i {
+			WithdrawAsset(..) => (),
+			_ => return Err(()),
+		}
```

Comment:
This could use `Option::filter` for brevity

```suggestion
		// Make sure the first instruction is DescendOrigin
		 iter
			.next()
			.filter(|instruction| matches!(instruction, DescendOrigin(_)))
			.ok_or(())?;

		// Then WithdrawAsset
		 iter
			.next()
			.filter(|instruction| matches!(instruction, WithdrawAsset(_)))
			.ok_or(())?;
```